### PR TITLE
[BEAM-9692] Make GroupByKey into a primitive

### PR DESCRIPTION
**Pull request by rohdesamuel**
This makes GBK into a primitive and moves the current implementation -- only used by the DirectRunner -- as a PTransformOverride for the DirectRunner. This leaves the apply_GroupByKey because there are some internal runners that rely on the existence of the transform. So we keep it but simply apply the GBK normally. This can be removed after an internal fix.
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by robertwb** [comment]
For this one, we should make GroupByKey primitive, and move the composite implementation to an override in the BundleBasedDirectRunner.

**Comment by rohdesamuel** [comment]
R: @robertwb

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 1]
```diff
@@ -234,7 +234,89 @@ def get_replacement_transform(self, transform):
...
+                        reify_output_type).with_output_types(gbk_input_type))
+                | (
+                    'GroupByWindow' >>
+                    _GroupAlsoByWindow(pcoll.windowing).with_input_types(
+                        gbk_input_type).with_output_types(gbk_output_type)))
+          else:
+            # The input_type is None, run the default
+            return (
+                pcoll
+                | 'ReifyWindows' >> ParDo(self.ReifyWindows())
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/direct/direct_runner.py:307]: Move ReifyWindows, _GroupByKeyOnly, _GroupAlsoByWindow, etc. here as well.

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 2]
```diff
@@ -234,7 +234,89 @@ def get_replacement_transform(self, transform):
...
+        self.input_type = applied_ptransform.inputs[0].element_type
+        return True
+      return False
+
+    def get_replacement_transform(self, ptransform):
+      # Imported here to avoid circular dependencies.
+      # pylint: disable=wrong-import-order, wrong-import-position
+      from apache_beam.transforms.core import GroupByKey
+
+      # Subclass from GroupByKey to inherit all the proper methods.
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/direct/direct_runner.py:257]: Hopefully there should be few (if any) methods that need to be inherited, which could simplify this.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py [hunk 3]
```diff
@@ -404,10 +396,6 @@ def test_gbk_then_flatten_input_visitor(self):
     flat = (none_str_pc, none_int_pc) | beam.Flatten()
     _ = flat | beam.GroupByKey()
 
-    # This may change if type inference changes, but we assert it here
-    # to make sure the check below is not vacuous.
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py:408]: Hmm... this check seems to be vacuous if the element type is already a tuple, right?

File: sdks/python/apache_beam/runners/direct/direct_runner.py (hunk 1 above)
Comment (rohdesamuel) [reply sdks/python/apache_beam/runners/direct/direct_runner.py]: Done

File: sdks/python/apache_beam/runners/direct/direct_runner.py (hunk 2 above)
Comment (rohdesamuel) [reply sdks/python/apache_beam/runners/direct/direct_runner.py]: Yep, I took out the subclass dependency and simplified the override

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py (hunk 3 above)
Comment (rohdesamuel) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner_test.py]: I added it back in. The DataflowRunner should change the element type and we should test before and after.

**Comment by rohdesamuel** [comment]
Thanks for the comments, I moved the GroupByKeyOnly etc to the DirectRunner and cleaned up the PTransformOverride by not subclassing.

**Comment by rohdesamuel** [comment]
Had to add back the ReifyWindows. Taking it out broke some internal tests.

**Comment by rohdesamuel** [comment]
Don't submit yet, there are some breaking internal tests

**Comment by rohdesamuel** [comment]
I added a condition to skip external transforms because we can assume that the external service provides correct transforms. Internal testing passing now.

**Pull request merged**